File: apps/TiflisCode/Shared/Domain/Models/ConnectionState.swift
//
//  ConnectionState.swift
//  TiflisCode
//
//  Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
//  Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
//

import Foundation
import SwiftUI

/// Represents the current connection state to the tunnel server
enum ConnectionState: Equatable {
    case disconnected
    case connecting
    case connected        // Tunnel connected, awaiting auth
    case authenticating   // Auth message sent, waiting for auth.success
    case authenticated    // Tunnel auth complete, heartbeat pending
    case verified         // End-to-end connectivity confirmed via heartbeat
    case degraded(String) // Heartbeat failing, connection may be stale
    case error(String)

    /// Returns true when authenticated or better (can send messages)
    var isConnected: Bool {
        switch self {
        case .authenticated, .verified, .degraded:
            return true
        default:
            return false
        }
    }

    /// Returns true only when end-to-end connectivity is verified via heartbeat
    var isVerified: Bool {
        if case .verified = self {
            return true
        }
        return false
    }

    /// Returns true if tunnel connection is established (but may not be authenticated yet)
    var isTunnelConnected: Bool {
        switch self {
        case .connected, .authenticating, .authenticated, .verified, .degraded:
            return true
        default:
            return false
        }
    }

    var indicatorColor: Color {
        switch self {
        case .verified:
            // Proper green (not lime/salad green)
            return Color(red: 0.2, green: 0.7, blue: 0.3)
        case .authenticated:
            // Yellow-green: tunnel auth ok, awaiting heartbeat verification
            return Color(red: 0.6, green: 0.8, blue: 0.2)
        case .degraded:
            return .orange
        case .connected, .authenticating, .connecting:
            return .yellow
        case .disconnected:
            return .gray
        case .error:
            return .red
        }
    }

    var indicatorSymbol: String {
        switch self {
        case .verified:
            return "●"
        case .authenticated, .degraded:
            return "●"
        case .connected, .authenticating, .connecting:
            return "◐"
        case .disconnected:
            return "○"
        case .error:
            return "●"
        }
    }

    var statusText: String {
        switch self {
        case .verified:
            return "Connected"
        case .authenticated:
            return "Verifying..."
        case .degraded(let reason):
            return "Degraded: \(reason)"
        case .authenticating:
            return "Authenticating..."
        case .connected:
            return "Connected to tunnel..."
        case .connecting:
            return "Connecting..."
        case .disconnected:
            return "Disconnected"
        case .error(let message):
            return "Error: \(message)"
        }
    }
}

File: apps/TiflisCodeAndroid/app/src/main/java/io/tiflis/code/domain/models/ConnectionState.kt
/*
 * Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
 * Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
 */

package io.tiflis.code.domain.models

/**
 * Represents the WebSocket connection state.
 * Uses sealed class pattern for type-safe state handling.
 */
sealed class ConnectionState {
    /** Not connected to any server */
    data object Disconnected : ConnectionState()

    /** Currently attempting to connect */
    data object Connecting : ConnectionState()

    /** Connected and authenticated with workstation */
    data object Connected : ConnectionState()

    /** End-to-end connection verified via heartbeat */
    data object Verified : ConnectionState()

    /** Connection degraded - heartbeat failing */
    data class Degraded(val reason: String) : ConnectionState()

    /** Connection lost, attempting to reconnect */
    data class Reconnecting(val attempt: Int) : ConnectionState()

    /** Connection error occurred */
    data class Error(val message: String, val code: String? = null) : ConnectionState()

    /** Whether we have an active authenticated connection (Connected, Verified, or Degraded) */
    val isConnected: Boolean
        get() = this is Connected || this is Verified || this is Degraded

    /** Whether connection has been verified end-to-end */
    val isVerified: Boolean
        get() = this is Verified

    val isConnecting: Boolean
        get() = this is Connecting || this is Reconnecting

    val displayText: String
        get() = when (this) {
            is Disconnected -> "Disconnected"
            is Connecting -> "Connecting…"
            is Connected -> "Authenticating…"
            is Verified -> "Connected"
            is Degraded -> "Connection Unstable"
            is Reconnecting -> "Reconnecting (attempt $attempt)…"
            is Error -> "Error: $message"
        }
}

/**
 * Connection credentials for tunnel server.
 */
data class ConnectionCredentials(
    val tunnelUrl: String,
    val tunnelId: String,
    val authKey: String
)

/**
 * Workstation information received after authentication.
 */
data class WorkstationInfo(
    val name: String?,
    val version: String?,
    val protocolVersion: String?,
    val workspacesRoot: String?
)

/**
 * Tunnel server information.
 */
data class TunnelInfo(
    val url: String,
    val id: String,
    val version: String?,
    val protocolVersion: String?
)
